### PR TITLE
Add test for deny-all network policy

### PR DESCRIPTION
**Pull request by sebrandon1**
build-depends: 25993

Depends on: https://github.com/test-network-function/cnf-certification-test-partner/pull/144

Adds a test to make sure that any network policies attached to namespaces running CNF workloads have a default `deny-all` type of network policy for both ingress and egress traffic.

This requires work on the side of the partner running their CNF workload to make sure they have not only a default network policy but a policy that allows the traffic that they need specifically.  Eg. allowing pods in a single namespace to talk to each other.

### CODE REVIEW

File: cnf-certification-test/networking/suite.go [hunk 1]
```diff
@@ -259,3 +266,58 @@ func testDualStackServices(env *provider.TestEnvironment) {
...
+						}
+					}
+				}
+			}
+		}
+
+		// Network policy has not been found that contains a deny-all rule for both ingress and egress.
+		if !denyAllIngressFound || !denyAllEgressFound {
+			namespacesMissingDenyAllDefaultPolicies = append(namespacesMissingDenyAllDefaultPolicies, put.Data.Namespace)
+			tnf.ClaimFilePrintf("Namespace %s was found to not have a default deny-all network policy.", put.Data.Namespace)
```
Comment (edcdavid) [inline cnf-certification-test/networking/suite.go:311]: Add which default policies are missing in the log, either ingress, egress or both

**Review by edcdavid — APPROVED** [review]

File: cnf-certification-test/networking/suite.go [hunk 2]
```diff
@@ -259,3 +266,58 @@ func testDualStackServices(env *provider.TestEnvironment) {
...
+			logrus.Debugf("Namespace: %s has already been processed for deny-all network policies.  Skipping.", put.Data.Namespace)
+			break
+		}
+
+		// Look through all of the network policies for a matching namespace.
+		for index := range env.NetworkPolicies {
+			// Match the pod namespace with the network policy namespace.
+			if put.Data.Namespace == env.NetworkPolicies[index].Namespace {
+				// Check to see if the network policy is deny-all (and contains ingress, egress, (or both) network policy types)
+				denyAllExists, foundPolicies := policies.IsNetworkPolicyDenyAll(&env.NetworkPolicies[index])
```
Comment (jmontesi) [inline cnf-certification-test/networking/suite.go:291]: I think that a more generic name (and more internal processing) for the function `IsNetworkPolicyDenyAll()` could simplify the code at this top level. For instance, something like `IsNetworkPolicyCompliant()` could return a `bool` and `error`, with the error indicating why it's not compliant. This way all the code to check the different requirements for a Network Policy to be compliant can be hidden inside the function and there is no need to return the policies.

File: internal/clientsholder/clientsholder.go [hunk 3]
```diff
@@ -37,19 +37,21 @@ import (
...
-	K8sClient     kubernetes.Interface
-	MachineCfg    ocpMachine.Interface
-	ready         bool
+	RestConfig          *rest.Config
+	DynamicClient       dynamic.Interface
+	APIExtClient        apiextv1.Interface
+	OlmClient           olmClient.Interface
+	OcpClient           clientconfigv1.ConfigV1Interface
+	K8sClient           kubernetes.Interface
+	K8sNetworkingClient networkingv1.NetworkingV1Interface
```
Comment (jmontesi) [inline internal/clientsholder/clientsholder.go:52]: This client can be gotten from the K8sClient (`clientsHolder.K8sClient.NetworkingV1()`).

File: cnf-certification-test/networking/suite.go [hunk 4]
```diff
@@ -259,3 +266,58 @@ func testDualStackServices(env *provider.TestEnvironment) {
...
 	}
 }
+
+//nolint:funlen
+func testNetworkPolicyDenyAll(env *provider.TestEnvironment) {
+	ginkgo.By("Test for Deny All in network policies")
+	var namespacesMissingDenyAllDefaultPolicies []string
+	var processedNamespaces []string
+
+	for _, put := range env.Pods {
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply cnf-certification-test/networking/suite.go:276]: Right but we only want to check namespaces that are used by CNF pods correct?

File: internal/clientsholder/clientsholder.go (hunk 3 above)
Comment (sebrandon1) [reply internal/clientsholder/clientsholder.go:52]: I was trying to make it consistent with how we use the normal `K8sClient` and passing around an interface type.  I tried getting a mock client up and mocking out all of the function calls to get the network policies but didn't have any luck.   The `networkingv1` import doesn't have the same fake client support it seems.

File: cnf-certification-test/networking/policies/policies.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+package policies
+
+import (
+	networkingv1 "k8s.io/api/networking/v1"
+)
+
+func IsNetworkPolicyDenyAll(np *networkingv1.NetworkPolicy) (bool, []networkingv1.PolicyType) {
+	if len(np.Spec.PodSelector.MatchLabels) != 0 {
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply cnf-certification-test/networking/policies/policies.go:24]: Good question.  I suppose there's more to be done if a pod selector is specified.

File: cnf-certification-test/networking/policies/policies.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+		return false, nil
+	}
+
+	// As long as we have decided above that there is no pod selector,
+	// we just have to make sure that the policy type is either Ingress or Egress (or both) we can return true.
+	// For more information about deny-all policies, there are some good examples on:
+	// https://kubernetes.io/docs/concepts/services-networking/network-policies/
+
+	if len(np.Spec.PolicyTypes) == 0 {
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply cnf-certification-test/networking/policies/policies.go:33]: I saw this as well.  I decided to try and prevent weird edge cases from slipping through and just made it explicit.

File: cnf-certification-test/networking/policies/policies.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+		return false, nil
+	}
+
+	// Ingress and Egress rules should be "empty" if it is a default rule.
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply cnf-certification-test/networking/policies/policies.go:37]: We will clarify the requirements document and see what we should do here.

**Comment by sebrandon1** [comment]
Marking this PR as work in progress again so we can clarify the requirements.

File: cnf-certification-test/networking/suite.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sebrandon1) [reply cnf-certification-test/networking/suite.go:337]: Switched my PR to look at network policies that have specific pod selector labels.

File: cnf-certification-test/networking/policies/policies.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (sebrandon1) [reply cnf-certification-test/networking/policies/policies.go:34]: Switched my PR to look at network policies that have specific pod selector labels.

File: cnf-certification-test/networking/policies/policies.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sebrandon1) [reply cnf-certification-test/networking/policies/policies.go]: Switched my PR to look at network policies that have specific pod selector labels.

File: cnf-certification-test/networking/suite.go (hunk 2 above)
Comment (sebrandon1) [reply cnf-certification-test/networking/suite.go:353]: Good idea, I'll try to improve.

**Comment by ramperher** [comment]
> * FAILURE https://www.distributed-ci.io/jobs/b2fcec72-1668-4628-8b03-2eb9e1085c8f/jobStates
>   from [Add test for deny-all network policy #423](https://github.com/test-network-function/cnf-certification-test/pull/423)

It is failing because we don't have default network policies created in our testing environment, I'll do it and test it together with this change to confirm it passes the test (it should). But don't worry about the warning, it is in our side :)

**Comment by sebrandon1** [comment]
@ramperher Yeah I figured once I finally get the CI fixed here I'd give a heads up to you prior to merging to add some network policies on your side for testing.  Just trying to get this thing to give me all of the green checks 😄

**Comment by ramperher** [comment]
> @ramperher Yeah I figured once I finally get the CI fixed here I'd give a heads up to you prior to merging to add some network policies on your side for testing. Just trying to get this thing to give me all of the green checks smile

I've created this to test the network policies, let's see! https://softwarefactory-project.io/r/c/dci-openshift-app-agent/+/25993

Then, I've put build-depends: 25993 in the description of this PR, so that DCI will take that code as base to run the debug job. I'll try a run right now

**Comment by ramperher** [comment]
> * FAILURE https://www.distributed-ci.io/jobs/9fa801d9-90a8-42b9-90a7-960519901edb/jobStates
>   from [Add test for deny-all network policy #423](https://github.com/test-network-function/cnf-certification-test/pull/423)

This job is from your last commit submitted, and we can see the test is passing. The job is marked as failed because it is failing the test related to ICMP connectivity, but it is my fault, I need to properly configure the network policies to allow ping between different namespaces. Apart from this, from my side it's fine to merge this change.

**Review by ramperher — APPROVED** [review]

**Comment by sebrandon1** [comment]
@ramperher For cross-namespace pinging, I have an example here: https://github.com/test-network-function/cnf-certification-test-partner/blob/main/test-target/pod-to-pod-np.yaml

**Comment by ramperher** [comment]
> @ramperher For cross-namespace pinging, I have an example here: https://github.com/test-network-function/cnf-certification-test-partner/blob/main/test-target/pod-to-pod-np.yaml

Yes, that's the one I was following yesterday, but I have two namespaces and several pods per namespace, I will continue checking this right now, anyway you can go on with this change, it's working fine. Thanks!

File: cnf-certification-test/networking/suite.go (hunk 1 above)
Comment (sebrandon1) [reply cnf-certification-test/networking/suite.go]: Completed.
